src/Integration/RuleSets/RuleSetManager.cs

```diff
@@ -0,0 +1,99 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+
+namespace SonarLint.VisualStudio.Integration.RuleSets
+{
+    internal sealed class RuleSetManager : IDisposable
+    {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly IProjectSystemHelper projectSystemHelper;
+        private readonly IRuleSetProvider rulesetProvider;
+
+        internal /* for testing purposes */ readonly ConcurrentDictionary<Language, SonarRuleSet> cachedRuleSets =
+            new ConcurrentDictionary<Language, SonarRuleSet>();
+
+        public RuleSetManager(IActiveSolutionBoundTracker activeSolutionBoundTracker, IProjectSystemHelper projectSystemHelper,
+            IRuleSetProvider rulesetProvider)
+        {
+            if (activeSolutionBoundTracker == null)
+            {
+                throw new ArgumentNullException(nameof(activeSolutionBoundTracker));
+            }
+
+            if (projectSystemHelper == null)
+            {
+                throw new ArgumentNullException(nameof(projectSystemHelper));
+            }
+
+            if (rulesetProvider == null)
+            {
+                throw new ArgumentNullException(nameof(rulesetProvider));
+            }
+
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.projectSystemHelper = projectSystemHelper;
+            this.rulesetProvider = rulesetProvider;
+
+            activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionBindingChanged;
+
+            if (activeSolutionBoundTracker.IsActiveSolutionBound)
+            {
+                FetchRuleSetForCurrentProjects();
+            }
+
+            // TODO: Find a way to listen to projects added to the solution (in case it is of a language we haven't yet fetched)
+        }
+
+        public void Dispose()
+        {
+            this.activeSolutionBoundTracker.SolutionBindingChanged -= OnSolutionBindingChanged;
+            cachedRuleSets.Clear();
+        }
+
+        private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventArgs e)
+        {
+            if (!e.IsBound)
+            {
+                cachedRuleSets.Clear();
+            }
+            else
+            {
+                FetchRuleSetForCurrentProjects();
+            }
+        }
+
+        private void FetchRuleSetForCurrentProjects()
+        {
+            this.projectSystemHelper.GetFilteredSolutionProjects()
```

Comment:
I am calling the `GetFilteredSolutionProjects` to not retrieve all projects from the solution but only the one with the right language + not excluded.

Reply:
I'm wondering whether it's worth filtering - why not just get the QPs for all of the supported languages?
Some projects e.g. ASP.Net could be multi-language (C# and JavaScript/TypeScript).
For now we could always just fetch C# and VB, if they exist.

Reply:
Good question... let's go for the non-filtered technique for now.